src/theory/datatypes/inference_manager.h

```diff
@@ -0,0 +1,111 @@
+/*********************                                                        */
+/*! \file inference_manager.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Datatypes inference manager
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__DATATYPES__INFERENCE_MANAGER_H
+#define CVC4__THEORY__DATATYPES__INFERENCE_MANAGER_H
+
+#include "context/cdhashmap.h"
+#include "expr/node.h"
+#include "theory/inference_manager_buffered.h"
+
+namespace CVC4 {
+namespace theory {
+namespace datatypes {
+
+/**
+ * The datatypes inference manager. The main unique features of this inference
+ * manager are:
+ * (1) Explicit caching of lemmas,
+ * (2) A custom process() method with relies on a policy determining which
+ * facts must be sent as lemmas (mustCommunicateFact).
+ * (3) Methods for tracking when lemmas and facts have been processed.
+ */
+class InferenceManager : public InferenceManagerBuffered
+{
+  typedef context::CDHashSet<Node, NodeHashFunction> NodeSet;
+
+ public:
+  InferenceManager(Theory& t, TheoryState& state, ProofNodeManager* pnm);
+  ~InferenceManager() {}
+  /**
+   * Reset, which resets flags regarding whether we have added lemmas or facts.
+   */
+  void reset();
+  /**
+   * Process the current lemmas and facts. This is a custom method that can
+   * be seen as overriding the behavior of calling both doPendingLemmas and
+   * doPendingFacts. It determines whether facts should be sent as lemmas
+   * or processed internally.
+   */
+  void process();
+  /**
+   * Have we asserted an internal fact (via process) since the most recent call
+   * to reset?
+   */
+  bool hasAddedFact() const;
+  /**
+   * Have we sent a lemma (via process) since the most recent call to reset?
+   */
+  bool hasAddedLemma() const;
+  /**
+   * If not cached, send lemma lem on the output channel and cache. Returns
+   * true if a lemma was sent.
+   */
+  bool doSendLemma(Node lem,
+                   LemmaProperty p = LemmaProperty::NONE,
+                   bool cached = true);
+  /** Multi-lemma version of the above, returns true if any lemma was sent. */
+  bool doSendLemmas(const std::vector<Node>& lem);
```

Comment:
Are these supposed to be `public`? Their doc should at least say something about their relation to `addPendingLemma()` and `lemma()`, when one should call which, etc.